framework/src/timeintegrators/AStableDirk4.C

```diff
@@ -95,6 +95,14 @@ AStableDirk4::computeTimeDerivatives()
   _du_dot_du = 1. / _dt;
 }
 
+void
+AStableDirk4::computeADTimeDerivatives(DualReal & ad_u_dot, const dof_id_type & dof)
```

Comment:
Are you sure there's no way to avoid copying this code?  A templated function maybe?

That goes for all of these time derivatives.

Reply:
I could create a templated helper in addition to the method already created that would cover the actual manipulation/calculation of the derivative. The method already created needs to exist in order to convert old solution etc. from a full NumericVector to a single value in order to have consistent operands for the AD case

Reply:
Mmm, I actually think this would be more work than it's worth. Take `BDF2`:
```c++
void
BDF2::computeTimeDerivatives()
{
  if (!_sys.solutionUDot())
    mooseError("BDF2: Time derivative of solution (`u_dot`) is not stored. Please set "
               "uDotRequested() to true in FEProblemBase befor requesting `u_dot`.");

  NumericVector<Number> & u_dot = *_sys.solutionUDot();
  if (_t_step == 1)
  {
    u_dot = *_solution;
    u_dot -= _solution_old;
    u_dot *= 1 / _dt;
    u_dot.close();

    _du_dot_du = 1.0 / _dt;
  }
  else
  {
    u_dot.zero();
    u_dot.add(_weight[0], *_solution);
    u_dot.add(_weight[1], _solution_old);
    u_dot.add(_weight[2], _solution_older);
    u_dot.scale(1. / _dt);
    u_dot.close();

    _du_dot_du = _weight[0] / _dt;
  }
}

void
BDF2::computeADTimeDerivatives(DualReal & ad_u_dot, const dof_id_type & dof)
{
  if (_t_step == 1)
  {
    const auto & local_old = _solution_old(dof);
    ad_u_dot -= local_old;
    ad_u_dot *= 1 / _dt;
  }
  else
  {
    ad_u_dot *= _weight[0];
    ad_u_dot += _weight[1] * _solution_old(dof);
    ad_u_dot += _weight[2] * _solution_older(dof);
    ad_u_dot *= 1. / _dt;
  }
}
```
We call `close` and `scale` on the `NumericVectors` which are methods simply not available to `Real/DualReals`. It looks like `NumericVector` does supply `+=` and `*=` overloads so those could be used in the helper...but then I'm going to have to factor out the `closes` and any math involving `_du_dot_du`...and there are 17 of these classes. 

It looks like it's possible if you want to make me do it.

Reply:
I don't know - maybe something for @permcody to decide.  He usually dislikes copies of code :-)

Reply:
Well none of us like copied code, at least I hope we don't. If there's not a better way, then just note that it's copied (list the source) and move on. Of course that still doesn't help if the original is modified since there's only a one way linkage of comments.

Reply:
@permcody you still didn't make a decision.  You're the decider...

Reply:
See https://github.com/idaholab/moose/pull/12738/commits/06fa9a8419a63ef464a37f71422ece844d0b9b06. That's about the best I can do for eliminating duplication